experimental/incremental/task.go

```diff
@@ -244,88 +369,115 @@ func (p *path) Walk() iter.Seq[*path] {
 
 // start executes a query in the context of some task and records the result by
 // calling done.
-func (t *task) start(caller Task, q *AnyQuery, done func(*result)) (async bool) {
+//
+// If sync is false, the computation will occur asynchronously. Returns whether
+// the computation is in fact executing asynchronously as a result.
+func (t *task) start(caller *Task, q *AnyQuery, sync bool, done func(*result)) (async bool) {
 	// Common case for cached values; no need to spawn a separate goroutine.
 	r := t.result.Load()
 	if r != nil && closed(r.done) {
 		done(r)
 		return false
 	}
 
+	if sync {
+		done(t.run(caller, q, false))
+		return false
+	}
+
 	// Complete the rest of the computation asynchronously.
 	go func() {
-		done(t.run(caller, q))
+		done(t.run(caller, q, true))
 	}()
 	return true
 }
 
 // run actually executes the query passed to start. It is called on its own
 // goroutine.
-func (t *task) run(caller Task, q *AnyQuery) (output *result) {
+func (t *task) run(caller *Task, q *AnyQuery, async bool) (output *result) {
 	output = t.result.Load()
 
 	defer func() {
-		if panicked := recover(); panicked != nil {
-			output = nil
-			caller.cancel(&ErrPanic{
-				Query:     q,
-				Panic:     panicked,
-				Backtrace: string(debug.Stack()),
-			})
+		if caller.aborted() == nil {
+			if panicked := recover(); panicked != nil {
+				output = nil
+				caller.cancel(&ErrPanic{
+					Query:     q,
+					Panic:     panicked,
+					Backtrace: string(debug.Stack()),
+				})
+			}
```

Comment:
It worries me a little that we don't have something like this in `Run`, to make sure panics don't inadvertently propagate, if they are done in the root goroutine. Though I think we could also prevent that by just not using panics for control flow.

Reply:
Well. This is only meant to catch panics coming out of `Query.Execute`. All other panics are the result of bugs, and should propagate out instead of being suppressed. So I think this design is correct.